### PR TITLE
Add group argument to server processes

### LINKED ISSUE #3178: Support balancing tablets using named tserver groups
Accumulo currently supports allocating tablets to specific servers using IP addresses and hostnames using certain balancer plugins. This provides a way to dedicate compute resources to specific tables.  Another way this could be accomplished is by adding a group option to tablet servers like [scan servers](https://github.com/apache/accumulo/blob/ab8e2f518887a2505cad9dbc02e82197babfc1bd/server/tserver/src/main/java/org/apache/accumulo/tserver/ScanServer.java#L130) have.  The group set for a tablet server could then be made available to balancer plugins.  Balancer plugins could use this information to assign specific tables or tablets to groups of tablet servers.

This feature would be useful for running groups of tablet servers in Kubernetes pods where the IP addresses may not be static.  Then tables could be assigned to those different groups giving different tables different qualities and quantities of dedicated compute resources.  It could also be used to simplify the balancer configuration on clusters with fixed ip addresses.

An example of how this could work is

 * on host A run `accumulo tserver -g meta`
 * on host B run `accumulo tserver -g meta`
 * on host C run `accumulo tserver`
 * on host D run `accumulo tserver`

After doing the above configure a balancer to assign the accumulo metadata table to group meta.  This would result in accumulo metadata tablets going to host A or B without having to configure the balancer with specifig information about host A and B.  The balancer plugin should assign all other tablets to host C and D.

To implement this feature would need to :
- Add the `-g` startup option to the tserver
- Store the group name in zookeeper for each tserver
- Modify the manager to get the group names from zookeeper and pass them to the balancer
- Possibly update the accumulo-cluster script to support group names for tservers in the cluster yaml file.

### PR DESCRIPTION
**Pull request by dlmarion**
This change elevates the group argument that was found in ScanServerOpts to ServerOpts, the options used as the base for all Accumulo server processes. The ScanServer used a different format for the lock data in ZooKeeper that comprised the host address and the group name, which allowed clients to find ScanServers that were part of a specific group. By moving the group from ScanServerOpts to ServerOpts it became necessary to embed the group name into the ZooKeeper lock data for all processes. Prior to this commit the ZooKeeper lock data was not using a defined structure, but a String that could be set to anything. I renamed the ServerServices class to ServiceLockData and created a well-defined structure for registering the Thrift Services that were exposed by the Accumulo server process to include the host, port, and group in this structure. Then, I modiified ServiceLock so that it expected and returned  a ServiceLockData object instead of a byte[].

Closes #3178

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java [hunk 1]
```diff
@@ -402,11 +404,10 @@ public Map<String,Pair<UUID,String>> getScanServers() {
       try {
         final var zLockPath = ServiceLock.path(root + "/" + addr);
         ZcStat stat = new ZcStat();
-        byte[] lockData = ServiceLock.getLockData(getZooCache(), zLockPath, stat);
-        if (lockData != null) {
-          String[] fields = new String(lockData, UTF_8).split(",", 2);
-          UUID uuid = UUID.fromString(fields[0]);
-          String group = fields[1];
+        ServiceLockData sld = ServiceLock.getLockData(getZooCache(), zLockPath, stat);
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:407]: I like the idea of creating the ServiceLockData to abstract away some of this implementation with more expressive APIs. I think that could be done independently of adding the group concept. Doing that first may make it easier to see just the changes required to add a tserver group feature, separately from the code organization work.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java [hunk 2]
```diff
@@ -512,24 +513,28 @@ public List<String> getManagerLocations() {
...
+          Thread.currentThread().getId(), zLockManagerPath);
       timer = new OpTimer().start();
     }
 
-    byte[] loc = zooCache.getLockData(zLockManagerPath);
+    ServiceLockData sld = zooCache.getLockData(zLockManagerPath);
+    String location = null;
+    if (sld != null) {
+      location = sld.getAddressString(ThriftService.MANAGER);
+    }
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:525]: You could make it so `sld` is never `null`, which would clean up some of these `null` checks, which seems to be a lot of boilerplate. You could return `Optional<ServiceLockData>` or a dummy `ServiceLockData` that represents the "not found in ZK" case (like `ServiceLockData.MISSING` that has a `null` address string).

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java [hunk 3]
```diff
@@ -512,24 +513,28 @@ public List<String> getManagerLocations() {
     OpTimer timer = null;
 
     if (log.isTraceEnabled()) {
-      log.trace("tid={} Looking up manager location in zookeeper.", Thread.currentThread().getId());
+      log.trace("tid={} Looking up manager location in zookeeper at {}.",
+          Thread.currentThread().getId(), zLockManagerPath);
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java:517]: Some of these log message changes seem unrelated to this PR. Might be useful to do those separately, first, so they don't detract from reviews on the main changes intended in this PR.

File: core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+     */
+    public static final String DEFAULT_GROUP_NAME = "default";
+
+    private final UUID uuid;
+    private final ThriftService service;
+    private final String address;
+    private final String group;
+
+    public ServiceDescriptor(UUID uuid, ThriftService service, String address, String group) {
+      super();
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java:69]: Pretty sure you don't need to explicitly call the super constructor here... especially since the super-class is just Object.
```suggestion
```

File: core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+      return toString().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o instanceof ServiceDescriptor) {
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java:99]: I think `instanceof` has gotten us in trouble before with `equals`. It will return true for subclasses, but the subclass may not have a reciprocal equals method implemented. Might want to instead check class same-ness rather than `instanceof`.

File: core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    }
+  }
+
+  private EnumMap<ThriftService,ServiceDescriptor> services;
+
+  public ServiceLockData(ServiceDescriptors sds) {
+    this.services = new EnumMap<>(ThriftService.class);
+    sds.getServices().forEach(sd -> {
+      this.services.put(sd.getService(), sd);
+    });
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java:150]:
```suggestion
    sds.getServices().forEach(sd -> this.services.put(sd.getService(), sd));
```

File: core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  }
+
+  public ServiceLockData(UUID uuid, String address, ThriftService service, String group) {
+    this(new ServiceDescriptors(new HashSet<>(
+        Collections.singleton(new ServiceDescriptor(uuid, service, address, group)))));
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java:155]:
```suggestion
    this(new ServiceDescriptors(Set.of(new ServiceDescriptor(uuid, service, address, group))));
```

File: minicluster/src/test/resources/log4j2-test.properties [hunk 8]
```diff
@@ -33,6 +33,6 @@ logger.01.level = error
 logger.02.name = org.apache.zookeeper
 logger.02.level = error
 
-rootLogger.level = info
+rootLogger.level = debug
```
Comment (ctubbsii) [inline minicluster/src/test/resources/log4j2-test.properties]: Is this temporary? I hope this doesn't make our tests during the build too spammy.

File: server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java [hunk 9]
```diff
@@ -27,6 +28,17 @@ public class ServerOpts extends ConfigOpts {
   @Parameter(names = {"-a", "--address"}, description = "address to bind to")
   private String address = null;
 
+  @Parameter(required = false, names = {"-g", "--group"},
+      description = "Optional group name that will be made available to the client (e.g. ScanServerSelector) "
+          + "and server (e.g. Balancers) plugins.  If not specified will be set to '"
+          + ServiceLockData.ServiceDescriptor.DEFAULT_GROUP_NAME
+          + "'. Assigning servers to groups support dedicating server resources for specific purposes, where supported.")
+  private String groupName = ServiceLockData.ServiceDescriptor.DEFAULT_GROUP_NAME;
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java:36]: I've never been a fan of these command-line options separate from our config file settings. The `-a` was added as a fix, but the semantics of it have always been confusing, as in... is it the "bind address" or is it the "service advertisement" address? It has always been conflated as both, with a special case for `0.0.0.0` and `localhost`, but it should always have been two.

I'm mentioning this now, because I don't think we should continue the precedent of adding new command-line options. Instead, there should be a `tserver.group` attribute that can be put in the config file or specified on the command-line, like any other config file option can be, with `-o tserver.group=X`. (Similarly, we should have `tserver.address.listen` and `tserver.address.advertise` as separate properties that default to `0.0.0.0` and `AUTO` by default, but that's a separate issue. My main point here is that we shouldn't continue the previous precedent... we should do it via config).

**Review by ctubbsii — CHANGES_REQUESTED** [review]
I like the changes to the ServiceLockData abstraction. I think that's a useful change on its own. After that change is done, I think the changes to add the service group are probably relatively minimal, so it's probably okay (meaning, I think I'd be in favor), but I'd like to see what this PR looks like after the ServiceLockData abstraction is done first, without the addition of the group. Would you be willing to do that?

File: core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java (hunk 5 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java:99]: This came over as part of the rename from ServerServices to ServiceLockData.

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java:99]: Gotcha. Still, it's something that we can fix now, since the rename effectively made it a new class, and it'd be good to fix it before any subclasses are created that would get affected by the change.

**Review by EdColeman — COMMENTED** [review]
This will likely make debugging test failures harder - why the change?  You can always set the level in your local copy, so the question becomes who is more inconvenienced by wanting to have the level at info vs debug.

**Comment by ctubbsii** [comment]
> This will likely make debugging test failures harder - why the change? You can always set the level in your local copy, so the question becomes who is more inconvenienced by wanting to have the level at info vs debug.

The current levels haven't been a problem at all, as far as I'm aware. This PR should probably just keep the same value that's there now so there's no risk of bloating the logs and spamming us or filling disks on CI servers. If there's a compelling reason to change it, we can definitely change it, but it hasn't been a problem to this point and it doesn't seem to be related to this PR (at least, it wasn't clear whether it was related). If Dave found it compelling to lower it to debug for testing this PR, or if we end up needing to do that a lot, that's probably a compelling reason to change it... as long as it's not too spammy. It just wasn't clear to me why it was being done here.

File: server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java (hunk 9 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java:36]: Does the current configuration mechanism support shell interpolation for configuration values? For example, could you put the following in `accumulo.properties`?
```
tserver.address.listen=$(hostname -i)`
```

If so, then this makes sense to me. If not, then I don't think anyone would want to put the bind address in a config file, as that would cause different `accumulo.properties` files on every host. At least by putting in on the command line you can use CLI commands to get the IP address and pass that in as the argument value. It even works in [k8s](https://github.com/dlmarion/accumulo-k8s/blob/main/accumulo-tserver.yaml#L23)

**Comment by dlmarion** [comment]
> I like the changes to the ServiceLockData abstraction. I think that's a useful change on its own. After that change is done, I think the changes to add the service group are probably relatively minimal, so it's probably okay (meaning, I think I'd be in favor), but I'd like to see what this PR looks like after the ServiceLockData abstraction is done first, without the addition of the group. Would you be willing to do that?

Yes, I would be willing to do that. However, if the ServiceLockData object doesn't support the [group](https://github.com/apache/accumulo/blob/main/server/tserver/src/main/java/org/apache/accumulo/tserver/ScanServer.java#L340), then that will break the [client](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java#L409) and likely break the build.

File: server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java [hunk 10]
```diff
@@ -27,6 +28,17 @@ public class ServerOpts extends ConfigOpts {
   @Parameter(names = {"-a", "--address"}, description = "address to bind to")
   private String address = null;
 
+  @Parameter(required = false, names = {"-g", "--group"},
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java:31]: Adding this option to server types where it does not do anything seems confusing to me.  Like I can start a manager with the option, but it will not have any effect.  Scoping it to only the tserver and scan server seems less confusing to me.

File: server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java (hunk 9 above)
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java:36]: We use commons-configuration2, which does support property interpolation, but not exactly like your example. You can do something like `tserver.address.listen=${env:HOSTNAME}` and in in your environment, you can do `HOSTNAME=$(hostname -i)` (for example, in `accumulo-env.sh` or `.bashrc`).

We also support setting any of our configuration properties on the command-line to override what's in the config file. So, instead of `-a $(hostname -i)`, you could do `-o tserver.address.listen=$(hostname -i)`, so you don't need to have the address hard-coded in the config file, and can set it dynamically on the command-line, even though it's a configuration file property. This feature was specifically added in 2.0 to support container deployments.

For the address, that would depend on us actually having a configuration property for the listen/bind address to replace `-a`, which we currently do not have (`tserver.address.listen` was merely an example). But, that's the direction I think we should go; for the purposes of this issue, I think the addition of the group should follow the direction of making it a configuration property, instead of the old precedent of `-a`. We have a complete configuration mechanism with a lot of flexibility already, so we shouldn't need to add one-off command-line options outside that configuration mechanism, like `-g` (and we should eventually get rid of the other one-off command-line options like `-a`, but that's for another ticket).

File: server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java (hunk 10 above)
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/ServerOpts.java:31]: My comment above regarding using our normal configuration property would make this a non-issue. Only server types where it makes sense would have that configuration property.

**Comment by ctubbsii** [comment]
> > I like the changes to the ServiceLockData abstraction. I think that's a useful change on its own. After that change is done, I think the changes to add the service group are probably relatively minimal, so it's probably okay (meaning, I think I'd be in favor), but I'd like to see what this PR looks like after the ServiceLockData abstraction is done first, without the addition of the group. Would you be willing to do that?
> 
> Yes, I would be willing to do that. However, if the ServiceLockData object doesn't support the [group](https://github.com/apache/accumulo/blob/main/server/tserver/src/main/java/org/apache/accumulo/tserver/ScanServer.java#L340), then that will break the [client](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java#L409) and likely break the build.

I am imagining the ScanServer's current `-g` option would go in the second PR, along with adding support for groups for tserver as well, so nothing is broken. I just think it makes sense to apply the "ServiceLockData abstraction/serialization" change feature first, then the "replace scan server group option with group added to ServiceLockData" feature after that, because they seem like they are discrete changes. I don't think the build needs to break to separate those out.

**Comment by dlmarion** [comment]
Considering the information about the configuration, I think I'll work on a PR to get rid of server process arguments, and replace them with properties. Then, once that is merged in, I can update this PR which should just be the ServiceLockData changes.

**Comment by dlmarion** [comment]
Created #3192 to remove arguments from the server processes

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java (hunk 2 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java]: Updated to use Optional

File: core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    }
+    return false;
+  }
+
+  @Override
+  public int compareTo(ServiceLockData other) {
+    return toString().compareTo(other.toString());
+  }
+
+  public static Optional<ServiceLockData> parse(String lockData) {
```
Comment (EdColeman) [inline core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java:217]: Would this be more general if it took `byte[]` and did the string conversion internally to the method?

**Comment by EdColeman** [comment]
I general this looks like a good change.  It may help if the internal representations of the lock data were more encapsulated so that the storage type was not visible outside of the LockData.

Looking at how to handle tservers and other processes that have locks when an upgrade occurs, one improvement could be if the lock data included either the data version or maybe the release version.  While added either of those at this point would be out of scope for this PR, designing it so that it is easier to extend could help limit changes in the future.

**Comment by dlmarion** [comment]
Full IT build passed

**Comment by dlmarion** [comment]
> It may help if the internal representations of the lock data were more encapsulated so that the storage type was not visible outside of the LockData

Can you be more specific about what you think should be changed? The storage type is a JSON string. The only serialization / deserialization that happens is inside of ServiceLockData.

**Comment by EdColeman** [comment]
What caught my eye was `ServiceLockData.parse(new String(data, UTF_8));` and was wondering if passing a `byte[]` would make it possible to isolate the internals more and then the rest was speculation on potential mods that could be able to use information stored within the lock itself.

**Review by EdColeman — APPROVED** [review]

**Comment by dlmarion** [comment]
Another Full IT build passed. I believe I resolved all issues. I'm going to merge this morning.

File: core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java (hunk 6 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java]: Implemented suggestion in 17df5d6.

File: core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java (hunk 5 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java]: Implemented suggestion in 17df5d6.

File: core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java (hunk 4 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/util/ServiceLockData.java]: Implemented suggestion in 17df5d6.
